contracts/near/src/lib.rs

```diff
@@ -0,0 +1,437 @@
+pub mod ext;
+pub mod state;
+
+const GAS_FOR_VERIFY_VAA: Gas = Gas(10_000_000_000_000);
+const STALENESS_THRESHOLD: u64 = 60;
+
+#[derive(BorshSerialize, BorshStorageKey)]
+enum StorageKeys {
+    Source,
+    Prices,
+}
+
+/// The `State` contains all persisted state for the contract. This includes runtime configuration.
+///
+/// There is no valid Default state for this contract, so we derive PanicOnDefault to force
+/// deployment using one of the #[init] functions in the impl below.
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct State {
+    /// The set of `Source`s from which Pyth attestations are allowed to be relayed from.
+    pub source: UnorderedSet<Source>,
+
+    /// A Mapping from PriceFeed ID to Price Info.
+    pub prices: UnorderedMap<Identifier, PriceFeed>,
+
+    /// The `AccountId` of the Wormhole contract used to verify VAA's.
+    pub wormhole: AccountId,
+
+    /// A hash of the current contract code.
+    ///
+    /// If this `hash` does not match the current contract code, then it indicates that a pending
+    /// upgrade has been allowed, calling `upgrade` with code that matches this hash will cause the
+    /// contract to upgrade itself.
+    pub codehash: [u8; 32],
+}
+
+/// The `Action` acts as an enum of possible `Action`'s this contract can receive via VAA.
+#[derive(BorshDeserialize, BorshSerialize)]
+pub enum Action {
+    AddSource(Source),
+    DelSource(Source),
+    BatchAttest(Vec<u8>),
+    UpgradeContract([u8; 32]),
+}
+
+#[near_bindgen]
+impl State {
+    #[init]
+    #[allow(clippy::new_without_default)]
+    pub fn new(wormhole: AccountId, codehash: [u8; 32]) -> Self {
+        Self {
+            source: UnorderedSet::new(StorageKeys::Source),
+            prices: UnorderedMap::new(StorageKeys::Prices),
+            wormhole,
+            codehash,
+        }
+    }
+
+    /// Add a new (chain, contract) pair to the list of trusted sources for Price Attestations.
+    #[private]
+    pub fn add_source(&mut self, source: Source) {
+        self.source.insert(&source);
+    }
+
+    /// Remove a (chain, contract) pair from the list of trusted sources for Price Attestations.
+    ///
+    /// NOTE: The caller of this function will be refunded the gas cost of the recovered storage
+    /// but the caller may not be the same as the original caller of `add_source`. This means the
+    /// refund goes to the wrong party. This is OK as the cost is low and the calls are rare and
+    /// the bookkeeping to fix this is not worth it.
+    #[private]
+    pub fn del_source(&mut self, source: Source) {
+        self.source.remove(&source);
+    }
+
+    pub fn get_sources(&self) -> Vec<Source> {
+        self.source.iter().collect()
+    }
+
+    /// Process a Batch of Price Attestations.
+    ///
+    /// This receives price updates, verifies them, and stores the attestations in the contract
+    /// state allowing other contracts to query them.
+    #[private]
```

Comment:
You can probably steal some docstrings from the Aptos contracts, which include links to the relevant GitBook pages.